server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/CommitIdDatabase.java

```diff
@@ -0,0 +1,264 @@
+            } while (buf.hasRemaining());
+
+            if (safeMode) {
+                channel.force(true);
+            }
+        } catch (IOException e) {
+            throw new StorageException("failed to update the commit ID database: " + path, e);
+        }
+
+        if (safeMode ||
+            headRevision == null ||
+            headRevision.major() < revision.major()) {
+            headRevision = revision;
+        }
+    }
+
+    void rebuild(Repository gitRepo) {
+        logger.warn("Rebuilding the commit ID database ..");
+
+        // Drop everything.
+        try {
+            channel.truncate(0);
+        } catch (IOException e) {
+            throw new StorageException("failed to drop the commit ID database: " + path, e);
+        }
+        headRevision = null;
+
+        // Get the commit IDs of all revisions.
+        try (RevWalk revWalk = new RevWalk(gitRepo)) {
+            final Revision headRevision;
+            final ObjectId headCommitId = gitRepo.resolve(R_HEADS_MASTER);
+            if (headCommitId == null) {
+                throw new StorageException("failed to determine the HEAD: " + gitRepo.getDirectory());
+            }
+
+            RevCommit revCommit = revWalk.parseCommit(headCommitId);
+            headRevision = CommitUtil.extractRevision(revCommit.getFullMessage());
+
+            // NB: We did not store the last commit ID until all commit IDs are stored,
+            //     so that the partially built database always has mismatching head revision.
+
+            ObjectId currentId;
+            Revision previousRevision = headRevision;
+            loop: for (;;) {
+                switch (revCommit.getParentCount()) {
+                    case 0:
+                        // End of the history
+                        break loop;
+                    case 1:
+                        currentId = revCommit.getParent(0);
+                        break;
+                    default:
+                        throw new StorageException("found more than one parent: " +
+                                                   gitRepo.getDirectory());
+                }
+
+                revCommit = revWalk.parseCommit(currentId);
+
+                final Revision currentRevision = CommitUtil.extractRevision(revCommit.getFullMessage());
+                final Revision expectedRevision = previousRevision.backward(1);
+                if (!currentRevision.equals(expectedRevision)) {
+                    throw new StorageException("mismatching revision: " + gitRepo.getDirectory() +
+                                               " (actual: " + currentRevision.major() +
+                                               ", expected: " + expectedRevision.major() + ')');
+                }
+
+                put(currentRevision, currentId, false);
+                previousRevision = currentRevision;
+            }
+
+            // All commit IDs but the head have been stored. Store the head finally.
```

Comment:
I feel that only the head is stored... how about just using except instead of but?

Reply:
Replaced but with except for clarity.